### PR TITLE
[turbine-llm] Add paged KV cache and llama variants to use it.

### PR DESCRIPTION
**Pull request by stellaraccident**
This is still just a reference implementation that linearizes in and out of the cache. I expect that the compiler will do some of this well enough with fusion that it will be fine, but we can replace hot areas with custom ops.

### CODE REVIEW

**Comment by stellaraccident** [comment]
@raikonenfnu This still needs some work/cleanup but is basically there in structure.

I need to do some more verification of numerics and expected results on Monday. Either this is a particularly dumb model or I accidentally dropped some IQ points in some math.

Example run:

```
python -m turbine_llm.inference.paged_llm_v1 ~/tmp/downloaded_open_llama_3b/ggml-model-f16.gguf ~/tmp/downloaded_open_llama_3b $'Q: What is calculus?\nA:' $'Q: What is the largest animal?\nA:' $'Q: Why is the sky black at night?\nA:' $'Q: How many band members were in the beatles?\nA:' $'Q: What are the names of the beatles and what instrument do they play?\nA:'
```

```
  0(21, 31): Calculus is the first used to be used to the the the the the the the the the the the
  1(4, 32): The blue whale.
  2(9, 34): Because the sky is a black at night.
  3(2, 36): 4
  4(17, 41): The Beatles are John Lennon, Paul George, George Harrison and Ringo Ring Ring.
```

**Comment by dan-garvey** [comment]
> @raikonenfnu This still needs some work/cleanup but is basically there in structure.
> 
> I need to do some more verification of numerics and expected results on Monday. Either this is a particularly dumb model or I accidentally dropped some IQ points in some math.
> 
> Example run:
> 
> ```
> python -m turbine_llm.inference.paged_llm_v1 ~/tmp/downloaded_open_llama_3b/ggml-model-f16.gguf ~/tmp/downloaded_open_llama_3b $'Q: What is calculus?\nA:' $'Q: What is the largest animal?\nA:' $'Q: Why is the sky black at night?\nA:' $'Q: How many band members were in the beatles?\nA:' $'Q: What are the names of the beatles?\nA:' $'Q: What instrument do each of the Beatles play?\nA:'```
> ```
> 
> 0(25, 35): Calculus is the first used to be used to the the the the the the the the the the the the the the the 1(4, 36): The blue whale. 2(9, 38): Because the sky is a black at night. 3(2, 40): 4 4(2, 39): The Beatles 5(21, 39): John Lennon plays the lead guitar, Paul plays the bass, George plays drums, plays the other instruments.

looking quantized

File: llm/turbine_llm/layers/kv_cache.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+            self.attn_head_count,
+            self.attn_head_dim,
+        ]
+        self.page_slab_flat_dim = math.prod(self.sub_page_dims)
+
+    def unflatten_page_table(self, state: list[torch.Tensor]) -> torch.Tensor:
+        """Unflattens the 2D page table to a 6D tensor."""
+        assert len(state) == 1, f"Expected 1-element state. Got: {len(state)}"
+        page_slab = state[0]
+        return page_slab.reshape(
```
Comment (raikonenfnu) [inline llm/turbine_llm/layers/kv_cache.py:73]: Probably NIT, but any thoughts on using view to guarantee no redundant copy?

File: llm/turbine_llm/layers/kv_cache.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        ]
+
+        # Reshape the page cache into sub-blocks so that we can index at the
+        # granularity of the transformer_block and cache partition.
+        # This requires us to recompute indices to the sub-block reference
+        # frame.
+        # The subblock slab is organized as:
+        #   [page, attn_layer, cache_partition]
+        # Where the cache line can be 0 (k) or 1 (v).
+        subblock_table = page_table.flatten(start_dim=0, end_dim=2)
```
Comment (raikonenfnu) [inline llm/turbine_llm/layers/kv_cache.py:132]: ditto Q wrt using view to guarantee no redundant copies.

File: llm/turbine_llm/layers/kv_cache.py (hunk 1 above)
Comment (stellaraccident) [reply llm/turbine_llm/layers/kv_cache.py:73]: I tend to use reshape unless if setting up for a mutation, which is the case that you need to be assured it is not a copy.

View imposes other costs for the guarantees it makes (around contiguity), which makes it a more tricky op in my view (no pun intended).

No free lunches...

Comment (raikonenfnu) [reply llm/turbine_llm/layers/kv_cache.py:73]: Thanks for the explanation, sg!

File: llm/turbine_llm/models/llama.py [hunk 3]
```diff
@@ -4,30 +4,72 @@
...
+################################################################################
+# Models
+################################################################################
+
+
+class PagedLlamaModelV1(BaseCausalLMModel):
+    """LlamaModel with a paged KV cache and supporting variable sequence
+    length batched inference.
 
-class DirectCacheLlamaModelV1(ThetaLayer):
```
Comment (raikonenfnu) [inline llm/turbine_llm/models/llama.py:25]: Any thoughts on leaving the naive implementation on for reference/education? :)

Comment (stellaraccident) [reply llm/turbine_llm/models/llama.py:25]: It's still there in the sibling file for that reason... Just wanted to have less to look at in this file

File: llm/turbine_llm/inference/paged_llm_v1.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    def release_page(self, index: int):
+        self.free_pages.append(index)
+
+
+class Batch:
+    def __init__(
+        self, parent: TorchGenerator, token_ids: torch.Tensor, seq_lens: torch.Tensor
+    ):
+        self.bs = token_ids.shape[0]
+        assert seq_lens.shape[0] == self.bs
```
Comment (raikonenfnu) [inline llm/turbine_llm/inference/paged_llm_v1.py:64]: nice :)

File: llm/scripts/validate_paged_llama_model.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    start_positions = torch.tensor([12, 6, 0, 0])
+    seq_lens = seq_lens + 1
+    decode_attention_mask = model.decode_attention_mask(
+        model.input_mask(
+            seq_lens,
+            seq_block_ids.shape[1] * model.cache.block_seq_stride,
+        ),
+        dtype=torch.float32,
+    )
+    logits = model.decode(
```
Comment (raikonenfnu) [inline llm/scripts/validate_paged_llama_model.py:117]: QQ: Where do we setup when to release certain sequences. For example what happen if once sequence is finished/found EOS before others?

File: llm/turbine_llm/models/llama.py (hunk 3 above)
Comment (raikonenfnu) [reply llm/turbine_llm/models/llama.py:25]: nice!

**Review by raikonenfnu — APPROVED** [review]
Went through the code, really cool stuff. 
Overall LGTM! :)

File: llm/turbine_llm/inference/paged_llm_v1.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+    def prefill(self):
+        model = self.parent.model
+        attention_mask = model.attention_mask(
+            model.input_mask(self.seq_lens, self.token_ids.shape[1]),
+            dtype=torch.float32,
+        )
+        seq_block_ids_tensor = self.pad_block_ids()
+        print(f":: Invoke prefill:")
+        trace_tensor("prefill.token_ids", self.token_ids)
```
[the remark this answers is not in the document]
Comment (stellaraccident) [reply llm/turbine_llm/inference/paged_llm_v1.py:123]: It's in a utility controlled currently by env var. It currently just prints but I was going to make it also be able to use the iree trace thing. So we can leave them in the code.

File: llm/scripts/validate_paged_llama_model.py (hunk 5 above)
Comment (stellaraccident) [reply llm/scripts/validate_paged_llama_model.py:117]: In this text, we don't. I'll end up deleting this test driver once the other one has more mileage on it. This was just the minimal I could do to test prefill and a single decode step.

In the real driver, there has to be an outer scheduler that assigns a sequence to a batch, and it keeps decoding until it observes it can stop (which may mean it overruns eos but the scheduler is responsible for ignoring).

Comment (raikonenfnu) [reply llm/scripts/validate_paged_llama_model.py:117]: sg!

File: llm/turbine_llm/inference/paged_llm_v1.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (raikonenfnu) [reply llm/turbine_llm/inference/paged_llm_v1.py:123]: sgsg
